ts/nni_manager/common/globals/shutdown.ts

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+/**
+ *  Shutdown manager.
+ *
+ *  Each module should register its clean up method with:
+ *
+ *      globals.shutdown.register('MyModuleName', async () => { this.cleanUp(); });
+ *
+ *  And a module can request for shutdown when unrecoverable error occurs:
+ *
+ *      try {
+ *          this.doSomethingMustSuccess();
+ *      } catch (error) {
+ *          globals.shutdown.criticalError('MyModuleName', error);
+ *      }
+ *
+ *  Note that when a module invokes `criticalError()`, its own registered callback will not get called.
+ *
+ *  When editting this module, keep robustness in mind.
+ *  Bugs in this module can easily swallow logs and make it difficult to reproduce users' issue.
+ **/
+
+import { Logger, getRobustLogger } from 'common/log';
+
+const logger: Logger = getRobustLogger('ShutdownManager');
+
+export class ShutdownManager {
+    private processStatus: 'initializing' | 'running' | 'stopping' = 'initializing';
+    private modules: Map<string, () => Promise<void>> = new Map();
+    private hasError: boolean = false;
+
+    public register(moduleName: string, shutdownCallback: () => Promise<void>): void {
+        if (this.modules.has(moduleName)) {
+            logger.error(`Module ${moduleName} has registered twice.`, new Error().stack);
+        }
+        this.modules.set(moduleName, shutdownCallback);
+    }
+
+    public initiate(reason: string): void {
+        if (this.processStatus === 'stopping') {
+            logger.warning('initiate() invoked but already stopping:', reason);
+        } else {
+            logger.info('Initiate shutdown:', reason);
+            this.shutdown();
+        }
+    }
+
+    public criticalError(moduleName: string, error: Error): void {
+        logger.critical(`Critical error ocurred in module ${moduleName}:`, error);
+        this.hasError = true;
+        if (this.processStatus === 'initializing') {
+            logger.error('Starting failed.');
+            process.exit(1);
+        } else if (this.processStatus !== 'stopping') {
+            this.modules.delete(moduleName);
+            this.shutdown();
+        }
+    }
+
+    public notifyInitializeComplete(): void {
+        if (this.processStatus === 'initializing') {
+            this.processStatus = 'running';
+        } else {
+            logger.error('notifyInitializeComplete() invoked in status', this.processStatus);
+        }
+    }
+
+    private shutdown(): void {
+        this.processStatus = 'stopping';
+
+        const promises = Array.from(this.modules).map(async ([moduleName, callback]) => {
```

Comment:
does the order of calling modules' callback matter?

Reply:
Good question.
The key is, if one module _owns_ another, the two modules should have composition relationship and only one of them should be exposed to public. For example `NNIManager` owns `NNIDataStore`, so `NNIDataStore` should be a private member of `NNIManager` and therefore should not be visible to any other modules, including `main.ts` and `globals.shutdown`. `NNIManager` is responsible for invoking `NNIDataStore`s constructor and destructor.
And if a module is a _global service_, like `LogStream`, it should be inside `globals` module. In this case, shutdown manager is logically aware of its existence and will destruct it after all callbacks.
This design does have limitations, as it does not support complex module relationships that cannot fit into above two cases. However, since NNI manager's task is relatively simple, it would be a sign of over-engineering if it required dependency resolving. So I decided to use a "flat" shutdown callback manager, rather than a more powerful "component manager".